configserver/src/main/java/com/yahoo/vespa/config/server/application/TenantApplications.java

```diff
@@ -88,52 +91,55 @@ public static TenantApplications create(Curator curator, ReloadHandler reloadHan
             catch (Exception e) {
                 log.log(LogLevel.WARNING, TenantRepository.logPre(tenant) + "Failed to clean up stray node '" + appNode + "'!", e);
             }
-            return Stream.empty();
+            return false;
         }
     }
 
+    /** Returns the id of the currently active session for the given application, if any. Throws on unknown applications. */
+    public OptionalLong activeSessionOf(ApplicationId id) {
+        String data = curator.getData(applicationPath(id)).map(Utf8::toString)
+                             .orElseThrow(() -> new IllegalArgumentException("Unknown application '" + id + "'."));
+        return data.isEmpty() ? OptionalLong.empty() : OptionalLong.of(Long.parseLong(data));
+    }
+
     /**
      * Returns a transaction which writes the given session id as the currently active for the given application.
      *
      * @param applicationId An {@link ApplicationId} that represents an active application.
      * @param sessionId Id of the session containing the application package for this id.
      */
-    public Transaction createPutApplicationTransaction(ApplicationId applicationId, long sessionId) {
-        if (listApplications().contains(applicationId)) {
+    public Transaction createPutTransaction(ApplicationId applicationId, long sessionId) {
+        if (curator.exists(applicationPath(applicationId))) {
             return new CuratorTransaction(curator).add(CuratorOperations.setData(applicationPath(applicationId).getAbsolute(), Utf8.toAsciiBytes(sessionId)));
         } else {
             return new CuratorTransaction(curator).add(CuratorOperations.create(applicationPath(applicationId).getAbsolute(), Utf8.toAsciiBytes(sessionId)));
         }
     }
 
     /**
-     * Return the stored session id for a given application.
+     * Return the active session id for a given application.
      *
      * @param  applicationId an {@link ApplicationId}
      * @return session id of given application id.
      * @throws IllegalArgumentException if the application does not exist
      */
-    public long getSessionIdForApplication(ApplicationId applicationId) {
-        String path = applicationPath(applicationId).getAbsolute();
-        try {
-            return Long.parseLong(Utf8.toString(curator.framework().getData().forPath(path)));
-        } catch (Exception e) {
-            throw new IllegalArgumentException(TenantRepository.logPre(applicationId) + "Unable to read the session id from '" + path + "'", e);
-        }
+    public long requireActiveSessionOf(ApplicationId applicationId) {
+        return activeSessionOf(applicationId)
+                .orElseThrow(() -> new IllegalArgumentException("Application '" + applicationId + "' has no active session."));
```

Comment:
Seems to me that there are users of this method that will not handle an exception being thrown, e.g. ApplicationRepository.lastDeployTime()

Reply:
This threw before, as well, when the application didn't have an active session. I'll check if that may have been unfortunate.